juice/src/layers/common/rnn.rs

```diff
@@ -67,29 +64,120 @@ pub enum RnnType {
+            "GRU" => Ok(RnnNetworkMode_UserInput::GRU),
+            "LSTM" => Ok(RnnNetworkMode_UserInput::LSTM),
+            "ReLU" => Ok(RnnNetworkMode_UserInput::ReLU),
+            "tanh" => Ok(RnnNetworkMode_UserInput::tanh),
             _ => Err("Unknown RnnType used - variants are GRU, LSTM, ReLU, and tanh"),
         }
     }
+
+    fn to_cudnn(&self) -> RnnNetworkMode {
+        match self {
+            RnnNetworkMode_UserInput::GRU => RnnNetworkMode::GRU,
+            RnnNetworkMode_UserInput::LSTM => RnnNetworkMode::LSTM,
+            RnnNetworkMode_UserInput::ReLU => RnnNetworkMode::ReLU,
+            RnnNetworkMode_UserInput::tanh => RnnNetworkMode::Tanh,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types)]
+/// Types of Supported RNN Input Modes
+pub enum RnnInputMode_UserInput {
+    /// Linear Input - A biased matrix multiplication is performed at the input of the first recurrent layer.
+    Linear,
+    /// Skip Input - No operation is performed to the input of the first recurrent layer
+    Skip,
+}
+
+impl RnnInputMode_UserInput {
+    fn to_text(&self) -> String {
+        match self {
+            RnnInputMode_UserInput::Skip => "Skip",
+            RnnInputMode_UserInput::Linear => "Linear",
+        }
+            .to_string()
+    }
+
+    fn from_text(input: &str) -> Result<Self, &str> {
+        match input {
+            "Linear" => Ok(RnnInputMode_UserInput::Linear),
+            "Skip" => Ok(RnnInputMode_UserInput::Skip),
+            _ => Err("Unknown RnnInputMode used - variants are Linear, Skip"),
+        }
+    }
+
+    fn to_cudnn(&self) -> RnnInputMode {
+        match self {
+            RnnInputMode_UserInput::Linear => RnnInputMode::LinearInput,
+            RnnInputMode_UserInput::Skip => RnnInputMode::SkipInput,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+#[allow(non_camel_case_types)]
+/// Types of Supported RNN DirectionModel
+pub enum DirectionMode_UserInput {
+    /// The network iterates recurrently from the first input to the last.
+    UniDirectional,
+    /// Each layer of the network iterates recurrently from the first input to the last and separately from the last input to the first. The outputs of the two are concatenated at each iteration giving the output of the layer.
+    BiDirectional,
+}
+
+impl DirectionMode_UserInput {
+    fn to_text(&self) -> String {
+        match self {
+            DirectionMode_UserInput::UniDirectional => "UniDirectional",
+            DirectionMode_UserInput::BiDirectional => "BiDirectional",
+        }
+            .to_string()
+    }
+
+    fn from_text(input: &str) -> Result<Self, &str> {
+        match input {
+            "UniDirectional" => Ok(DirectionMode_UserInput::UniDirectional),
+            "BiDirectional" => Ok(DirectionMode_UserInput::BiDirectional),
+            _ => Err("Unknown DirectionMode used - variants are UniDirectional, BiDirectional"),
+        }
+    }
+
+    fn to_cudnn(&self) -> DirectionMode {
```

Comment:
it's not really cudnn, it's backend agnostic, but I think this should go away anyways or use the `TryFrom`